app/Services/ModuleService.php

```diff
@@ -70,4 +81,139 @@ public function getAdminLinks(): array
     {
         return self::$adminLinks;
     }
+
+    public function createModule($array): bool
+    {
+        $orig_file = $array['file'];
+        $file_ext = $orig_file->getClientOriginalExtension();
+        $allowed_extensions = ['zip', 'tar', 'gz'];
+
+        if (!in_array($file_ext, $allowed_extensions)) {
+            return false;
+        }
+
+        $module = null;
+        $temp = storage_path('app/tmp');
+        $zipper = null;
+
+        if ($file_ext === 'tar' || $file_ext === 'gz') {
+            $zipper = new PharData($orig_file);
+            $zipper->decompress();
+        }
+
+        if ($file_ext === 'zip') {
+            $madZipper = new Madzipper();
+
+            try {
+                $zipper = $madZipper->make($orig_file);
+            } catch (Exception $e) {
+                Log::emergency('Could not extract zip file.');
+            }
+        }
+
+        try {
+            $zipper->extractTo($temp);
+        } catch (Exception $e) {
+            Log::emergency('Cannot Extract Module!');
+        }
+
+        if (!File::exists($temp.'/module.json')) {
+            $directories = Storage::directories('tmp');
+            $temp = storage_path('app').'/'.$directories[0];
+        }
+
+        $file = $temp.'/module.json';
+
+        if (File::exists($file)) {
+            $json = json_decode(file_get_contents($file), true);
+            $module = $json['name'];
+        } else {
+            return false;
+        }
+
+        if (!$module) {
+            return false;
+        }
+
+        $toCopy = base_path().'/modules/'.$module;
+
+        if (File::exists($toCopy)) {
+            return false;
+        }
+
+        File::moveDirectory($temp, $toCopy);
+
+        $this->addModule($module, $array['enabled']);
+
+        Artisan::call('config:cache');
+        return true;
+    }
+
+    public function updateModule($id, $status): bool
+    {
+        $module = (new Module())->find($id);
+        $module->update([
+            'enabled' => $status,
+        ]);
+        return true;
+    }
+
+    public function deleteModule($id, $data): bool
+    {
+        $module = (new Module())->find($id);
+        if ($data['verify'] === strtoupper($module->name)) {
+            try {
+                $module->delete();
+            } catch (Exception $e) {
+                Log::emergency('Cannot Delete Module!');
+            }
+            $moduleDir = base_path().'/modules/'.$module->name;
+            if (File::exists($moduleDir)) {
+                File::deleteDirectory($moduleDir);
```

Comment:
Also try-catch around this to check that the directory already exists. The user might have deleted it manually but is now removing it from the admin panel this way